src/std/actor-v18/connection.ss

```diff
@@ -149,71 +153,74 @@
+          (spawn/name 'actor-connection actor-connection srv peer-id sock reader writer 'out)
+          (begin
+            (warnf "peer id mismatch for TLS authenticated peer ~a" (peer-address sock))
+            (with-catch void (cut sock.close))
+            'error)))
+      ;; no TLS; do cookie authentication handshake
+      (let/cc exit
+        (def (fail! what)
+          (warnf "handshake with ~a failed: ~a" peer-id what)
+          (with-catch void (cut sock.close))
+          (thread-send/check srv (!connection-failed peer-id what))
+          (exit 'error))
 
-      (try
-       (let* ((reader (StreamSocket-reader sock))
-              (reader (open-buffered-reader reader))
-              (writer (StreamSocket-writer sock))
-              (writer (open-buffered-writer writer))
-              (srv-id (thread-specific srv)))
-         ;; set handshake timeouts
-         (let (expiry (timeout->expiry default-handshake-timeout))
-           (StreamSocket-set-input-timeout! sock expiry)
-           (StreamSocket-set-output-timeout! sock expiry))
-         ;; send hello
-         (&BufferedWriter-write-u16 writer version-magic)
-         (write-delimited writer (!hello srv-id))
-         (when (eof-object? (&BufferedReader-peek-u8 reader))
-           (fail! "incomplete handshake"))
-         (match (read-delimited reader)
-           ((!challenge remote-id salt)
-            (cond
-             ((not remote-id)
-              (fail! "bad challenge; no server id"))
-             ((eq? remote-id srv-id)
-              (fail! "bad challenge; server claims to be our server"))
-             ((not (eq? remote-id peer-id))
-              (fail! "bad challenge; server id mismatch"))
-             (else
-              (let (cli-salt (random-bytes (u8vector-length cookie)))
-                (write-delimited writer (!response (digest peer-id srv-id salt cookie) cli-salt))
-                (match (read-delimited reader)
-                  ((!accepted auth)
-                   (unless (equal? auth (digest peer-id srv-id cli-salt cookie))
-                     (fail! "server authentication failed"))
-                   (infof "connected to server: ~a" peer-id)
-                   ;; reset timeouts
-                   (StreamSocket-set-input-timeout! sock #f)
-                   (StreamSocket-set-output-timeout! sock #f)
-                   ;; and go!
-                   (spawn/name 'actor-connection actor-connection srv peer-id sock reader writer 'out))
-                  ((? eof-object?)
-                   (fail! "incomplete handshake"))
-                  (unexpected
-                   (fail! "bad response")))))))
-           ((? eof-object?)
-            (fail! "incomplete handshake"))))
-       (catch (exn)
-         (warnf "unhandled exception: ~a" exn)
-         (fail! (error-message exn)))))))
+        (try
+         (let* ((reader (sock.reader))
+                (reader (open-buffered-reader reader))
+                (writer (sock.writer))
```

Comment:
I don't like this re-binding or reader. How about instead a single-line `(open-buffered-reader (sock.reader))` ?
Same in other parts.

Reply:
I do that a lot and it leads to easier to read code. That's one of the prime uses of let*.
I don't like to do too many things in the same line, I am dealing with tons of code and my eyes get strained.
So sparser code, get used to it! And I also recommend you adopt this style instead of putting underscores,.